core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1721,15 +1718,49 @@ object Stream {
      * because `s1.interruptWhen(s2)` is never pulled for another element after the first element has been
      * emitted. To fix, consider `s.flatMap(_ => infiniteStream).interruptWhen(s2)`.
      */
-    def interruptWhen(haltWhenTrue: Stream[F,Boolean])(implicit F: Effect[F], ec: ExecutionContext): Stream[F,O] =
-      haltWhenTrue.noneTerminate.either(self.noneTerminate).
-        takeWhile(_.fold(halt => halt.map(!_).getOrElse(false), o => o.isDefined)).
-        collect { case Right(Some(i)) => i }
+    def interruptWhen(haltWhenTrue: Stream[F,Boolean])(implicit F: Effect[F], ec: ExecutionContext): Stream[F,O] = {
```

Comment:
Do we need a similar refactoring for `pauseWhen`?

Reply:
I would like to go over all nondet combinators when that is done and clean them up. I think we may need to touch mergeXX as well. But prefferably I would like to do this as separate pr ?